core/src/main/scala/com/softwaremill/react/kafka_api_opt3/Consumer.scala

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt3
+
+import scala.collection.immutable
+import scala.concurrent.duration.FiniteDuration
+import akka.stream.Inlet
+import org.apache.kafka.common.TopicPartition
+import akka.stream.stage.GraphStageWithMaterializedValue
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import akka.stream.Outlet
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import akka.stream.SourceShape
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import akka.stream.Shape
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import akka.stream.scaladsl.Flow
+import scala.concurrent.Future
+import akka.NotUsed
+import akka.Done
+
+object Consumer {
+
+  trait Message[K, V] {
+    def transaction: Transaction
+    def record: ConsumerRecord[K, V]
+  }
+
+  trait Transaction {
+    def commit(): Future[Done]
```

Comment:
I would say that unless explicitly overridden (with a possible configuration flag) commits should be gapless, i.e. similar semantics to mapAsync:

 - M4 `commit()` called
 - M2 `commit()` called
 - M1 `commit()` called
 - M1 M2 is committed to Kafka
 - M3 `commit()` called
 - M3 M4 committed to Kafka

Btw, how does the current implementation handle this? Also, this is not specific to Option 3, Option 1 uses an explicit channel not a hidden side-channel but the problems are exactly the same.